### PR TITLE
Add a minimal `spack update` command

### LINKED ISSUE #9134: Add a `spack update` command that updates a user's local Spack git repo
### `spack update`: like `brew update`, but for `spack`

A command that behaves like Linuxbrew/Homebrew's `brew update`. That is, it:

- fetches `develop` from `origin`
- merges fetched `develop` where appropriate
- possibly stashes existing changes on current branch, checks out `develop`, pulls `develop` (i.e., `git-fetch` and `git-merge`), checks out the old current branch, and re-applies stashed changes
- does not install or build anything
- does not do anything more than fetch/merge/stash/unstash operations on the local git repo copy of someone's `spack` installation

### Rationale

Right now, to update `spack`, I need to

```
pushd ${SPACK_ROOT}
git checkout develop && git fetch origin && git merge --ff-only origin/develop && \ 
git checkout $MY_OLD_BRANCH # or use git-pull, or simiilar
popd
```

whereas to do something vaguely equivalent in Homebrew is `brew update`. Consequently, it's way easier to stay up-to-date with `homebrew-core/master` than it is to stay up-to-date with `spack`.

### Description

See above for longer description. Succinctly, I'd like a `spack-update` command that behaves like `brew-update` does for Homebrew/Linuxbrew.

### Alternatives

- Do nothing. Fine by me, but the number of commands required slows me down in comparison to running a single command to update my copy of spack's github repo.
- Have `spack update` do more stuff, like actually upgrade packages, potentially even in-place, or install new packages; I think adding more functionality to the general idea of "implement a single command that just keeps the spack repo up-to-date and nothing else" starts into feature creep.

### PR DESCRIPTION
**Pull request by alecbcs**
This PR adds a minimal `spack update` command which mimics the functionality of similar package managers (e.g. `brew`, `apt`, etc...) allowing users to update their Spack installation without requiring them to explicitly use git in their spack installation directory. *This command is not aimed at upgrading packages.*

### Demo Output:
**Already up to date.**
```
alec@laptop spack % spack update
Already up to date.
```

**Help new users understand the consequences of different branches.**
```
alec@laptop spack % spack update
==> Warning: Spack is not tracking upstream develop.
==> Warning: Packages may be out of date. To switch to develop run,

    spack update -b develop

Already up to date.
```

**Improved human readable `git pull` results.**
```
alec@laptop spack % spack update
Updating 0d37564a..eacb8242
Fast-forward
==> Added 2 packages
perl-email-simple  perl-moox-types-mooselike
==> Updated 4 packages
berkeleygw  fmt  gperftools  py-tensorboard
```

### Design Goals:
- Make use of git plumbing instead of porcelain commands whenever parsing git output.
- Show standard `git pull` output to users instead of trying to overwrite it with our own.
- This command is not a replacement for `git`. If you're actively developing on Spack you should use normal git. This command provides a nice interface for a common action for Spack new/casual users.

Fixes #9134.

### CODE REVIEW

**Comment by haampie** [comment]
I would reserve `update` for environments (not sure what it would mean, but something like `concretize --force && install`). Maybe `spack self-update`?

**Comment by teojgo** [comment]
> I would reserve `update` for environments (not sure what it would mean, but something like `concretize --force && install`). Maybe `spack self-update`?

Mabe `spack sync` could be a candidate.

**Comment by alecbcs** [comment]
> I would reserve `update` for environments (not sure what it would mean, but something like `concretize --force && install`). Maybe `spack self-update`?

What you're describing here is what I'd call `spack upgrade`. That's also on my todo list as kinda the part two to this part one. I'd argue for this command being update and that one being upgrade so that we can build off of the similar syntax used by others like brew and apt. Although, I'm open to a debate if others dislike the syntax used by other package managers.

**Comment by adamjstewart** [comment]
The update/upgrade difference in brew always confused me, but if we have to use those names, we should definitely match them. I get the same complaint about list/find from a lot of users. This also relates to #4159.

**Comment by tgamblin** [comment]
@adamjstewart this does mirror the semantics from `brew`

**Comment by alecbcs** [comment]
> Mabe `spack sync` could be a candidate.

@adamjstewart @haampie @tgamblin what do you think of `sync` as suggested by from @teojgo?

**Comment by alecbcs** [comment]
@alalazo are we removing this from the v0.21.0 roadmap?

**Comment by alalazo** [comment]
@alecbcs For the last release we had a lot of issues / PRs in the milestone, and most of them were pruned as we approached the release date due to lack of time. This was both time consuming for maintainers, and I think confusing for users. 

This time, we are trying to keep in the release milestone only the features and issues that we are confident we can have, so people can have a better view of what is going on. This is also in line with adding a [Known bugs](https://github.com/orgs/spack/projects/6) project, where people can check which bugs are being fixed / have been fixed recently.

**TL;DR** This is out for now, but we can add it back to the milestone for v0.21.0 when close to getting merged.

**Comment by alecbcs** [comment]
@adamjstewart @haampie @alalazo do you all have any reservations about this PR? 

With @tgamblin's proposal on `spack upgrade` I think a `spack update` makes a lot of sense as a natural complement to mirror syntax/functionality from other much larger package managers such as `apt` and `brew`.

**Review by adamjstewart — DISMISSED** [review]
Haven't actually tested, but I approve of this in principle. Would be nice to add type hints to functions but we barely have those in Spack at the moment anyway. Wonder if there's a coverage tool for type hints...

**Comment by alecbcs** [comment]
I've been adding type hints to library functions as I create them, but have so far held off on doing so for command definitions since I didn't see any others implementing them and wasn't exactly sure how to work with the subparser library for typing.

In general though I totally agree, we should start adding more type hints wherever possible.

**Comment by adamjstewart** [comment]
I added type hints to the `commands` command if you need an example.

**Comment by alecbcs** [comment]
@adamjstewart thanks for the pointer! That actually helped a lot.

**Review by adamjstewart — DISMISSED** [review]

File: lib/spack/spack/cmd/update.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import spack.util.git
+from spack.cmd import spack_is_git_repo
+
+description = "update the spack repository to the latest commit"
+section = "system"
+level = "short"
+
+
+def setup_parser(subparser: ArgumentParser) -> None:
+    subparser.add_argument("-b", "--branch", help="name of the branch to upate the repository to")
```
Comment (tgamblin) [inline lib/spack/spack/cmd/update.py:25]:
```suggestion
    subparser.add_argument("-b", "--branch", help="name of a branch to update the repository to")
```

but I think this would be better as:

```suggestion
    mutex = subparser.add_mutually_exclusive_group()
    mutex.add_argument("release_or_branch", nargs="?", "--help="name of a branch or release to use (optional)")
    
    mutex.add_argument("-l", "--list-releases", "--help="list names of known spack releases")
    mutex.add_argument("--list-branches", "--help="list names of all known branches")
```

where:
* `spack update` pulls the current branch
* `spack update <ref>` checks out the given release or branch (they're the same but I think we should just show releases by default)
* and the two list options list either just release branches or all branches

File: lib/spack/spack/cmd/update.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+
+def update(parser: ArgumentParser, args: Namespace) -> None:
+    # make sure that spack is a git repository
+    if not spack_is_git_repo():
+        tty.die("This spack is not a git clone. Can't use 'spack update'")
+
+    git = spack.util.git.git()
+    if git is None:
+        tty.die("Couldn't find a usable git executable in PATH. Unable to update.")
```
Comment (tgamblin) [inline lib/spack/spack/cmd/update.py:35]: for a future PR you *could* make it set up a git repo if we're using a tarball release, but not really necessary for the initial PR.

File: lib/spack/spack/cmd/update.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                branch = ref.replace("refs/heads/", "", 1)
+                branch_remote_ref = git("config", f"branch.{branch}.merge", output=str)
+
+                if branch_remote_ref.strip().endswith("develop"):
+                    remote = git("config", f"branch.{branch}.remote", output=str).strip()
+                    remote_url = git("config", f"remote.{remote}.url", output=str).strip()
+
+                    if re.match(r"spack\/spack(\.git)?$", remote_url):
+                        upstream_branch = branch
+                        break
```
Comment (tgamblin) [inline lib/spack/spack/cmd/update.py:58]: split the branch finding out into a util function and add it to `spack.util.git`

File: lib/spack/spack/cmd/update.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+            if upstream_branch is not None and current_branch != upstream_branch:
+                tty.warn("Spack is not tracking upstream develop.")
```
Comment (tgamblin) [inline lib/spack/spack/cmd/update.py:61]: this should keep in mind the structure of our [release branches](https://github.com/spack/spack/#releases) and allow the user to `spack update` point releases on a release branch.

I don't think it should prompt you to switch to develop if you are on a release, and it should probably say something like "you're on v0.21.0 and 0.21.3 is available. updating to 0.21.3" if you are on a release branch, as this is how we encourage people to use them.

**Review by tgamblin — CHANGES_REQUESTED** [review]

File: lib/spack/spack/cmd/update.py (hunk 2 above)
Comment (alecbcs) [reply lib/spack/spack/cmd/update.py:35]: Would your preference be for setting up a git repository over simply downloading the latest tarball and unpacking it over top of the existing installation?

File: lib/spack/spack/cmd/update.py (hunk 1 above)
Comment (alecbcs) [reply lib/spack/spack/cmd/update.py:25]: I'm game for `release_or_branch` and `-l` but I'm not sure how useful `--list-branches` will be unless we stop encouraging folks from creating branches directly in the Spack repository as we now have ~125 branches. I think if you're looking for a branch outside of `develop` and releases you're more likely to know about and use `git branch`.
